### PR TITLE
ICU-21349 Add UnitsConverter.getConversionInfo()

### PR DESCRIPTION
**Pull request by younies**
Closes: https://github.com/icu-units/icu/issues/24

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21349
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/test/intltest/units_test.cpp [hunk 1]
```diff
@@ -173,6 +175,89 @@ void UnitsTest::testExtractConvertibility() {
...
+        {
+            "fahrenheit",
+            "celsius",
+            {5.0 / 9.0, -17.7777777778, false},
+        },
+        {
+            "celsius",
+            "fahrenheit",
+            {9.0 / 5.0, 32, false},
+        },
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:220]:
```suggestion
        },
        {
            "fahrenheit",
            "fahrenheit",
            {1, 0, false},
        },
```

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/impl/UnitsTest.java [hunk 2]
```diff
@@ -271,6 +271,89 @@ public void testExtractConvertibility() {
...
+                        "kelvin",
+                        5.0 / 9.0, 255.372, false),
+                new TestData(
+                        "fahrenheit",
+                        "celsius",
+                        5.0 / 9.0, -17.7777777778, false),
+                new TestData(
+                        "celsius",
+                        "fahrenheit",
+                        9.0 / 5.0, 32, false),
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/impl/UnitsTest.java:318]:
```suggestion
                new TestData(
                        "celsius",
                        "fahrenheit",
                        9.0 / 5.0, 32, false),
                new TestData(
                        "fahrenheit",
                        "fahrenheit",
                        1.0, 0, false),
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 3]
```diff
@@ -173,6 +175,89 @@ void UnitsTest::testExtractConvertibility() {
...
+            maxDelta = 1e-12;
+        }
+        assertEqualsNear(message + " for conversion rate: ", testCase.expectedConversionInfo.conversionRate,
+                         actualConversionInfo.conversionRate, maxDelta);
+
+        maxDelta = 1e-6 * uprv_fabs(testCase.expectedConversionInfo.offset);
+        if (testCase.expectedConversionInfo.offset == 0) {
+            maxDelta = 1e-12;
+        }
+        assertEqualsNear(message + " for offset: ", testCase.expectedConversionInfo.offset, actualConversionInfo.offset,
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:253]:
```suggestion
        assertEqualsNear(message + ", offset: ", testCase.expectedConversionInfo.offset, actualConversionInfo.offset,
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 4]
```diff
@@ -173,6 +175,89 @@ void UnitsTest::testExtractConvertibility() {
...
+                         maxDelta);
+
+        assertEquals(message + "for reciprocal: ", testCase.expectedConversionInfo.reciprocal,
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:256]:
```suggestion
        assertEquals(message + ", reciprocal: ", testCase.expectedConversionInfo.reciprocal,
```

**Review by hugovdm — COMMENTED** [review]
(Review at 8446cda)

Looks good - just a couple of comments, with suggestions: if you like the suggestions, you can apply them and squash immediately.

**Comment by hugovdm** [comment]
I've also changed the PR description, since "extract conversion info" made me wonder "to where?"
(I think my suggested PR title would be clearer for readers of the git commit log.)

File: icu4c/source/i18n/units_converter.h [hunk 5]
```diff
@@ -82,6 +82,12 @@ struct U_I18N_API Factor {
     void substituteConstants();
 };
 
+struct U_I18N_API ConversionInfo {
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/units_converter.h:85]: Acknowledges :)

File: icu4c/source/test/intltest/units_test.cpp [hunk 6]
```diff
@@ -173,6 +175,89 @@ void UnitsTest::testExtractConvertibility() {
...
+
+        double maxDelta = 1e-6 * uprv_fabs(testCase.expectedConversionInfo.conversionRate);
+        if (testCase.expectedConversionInfo.conversionRate == 0) {
+            maxDelta = 1e-12;
+        }
+        assertEqualsNear(message + " for conversion rate: ", testCase.expectedConversionInfo.conversionRate,
+                         actualConversionInfo.conversionRate, maxDelta);
+
+        maxDelta = 1e-6 * uprv_fabs(testCase.expectedConversionInfo.offset);
+        if (testCase.expectedConversionInfo.offset == 0) {
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/test/intltest/units_test.cpp:250]: done.

File: icu4c/source/test/intltest/units_test.cpp (hunk 1 above)
Comment (younies) [reply icu4c/source/test/intltest/units_test.cpp:220]: I have added it

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/impl/UnitsTest.java (hunk 2 above)
Comment (younies) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/impl/UnitsTest.java:318]: I have added it

**Review by hugovdm — APPROVED** [review]
Looks good!
(at fc94025fa60e5a707fa0128381cdde0450223c9a)

**Review by hugovdm — APPROVED** [review]

**Pull request merged**